File: hard/str_substitution.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    chomp $line;
    my ( $l, $r ) = split( /;/, $line );
    if ( $r eq '' ) {
        print "\n";
        next;
    }
    my @s = split( /,/, $r );
    for ( my $i = 0 ; $i < scalar @s ; $i += 2 ) {
        if ( $s[ $i + 1 ] eq '' ) {
            $s[ $i + 1 ] = "x";
        }
        else {
            $s[ $i + 1 ] =~ tr/01/ab/;
        }
        $l =~ s/$s[$i]/$s[$i + 1]/g;
    }
    $l =~ tr/abx/01/d;
    printf "$l\n";
}
close(INFILE);

File: hard/str_substitution.rb
File.open(ARGV[0]).each_line do |line|
  s = line.chomp.split(';')
  if s.length < 2
    puts
    next
  end
  t = s[1].split(',')
  0.step(t.length - 1, 2) do |i|
    if t[i + 1].nil? || t[i + 1].empty?
      t[i + 1] = 'x'
    else
      t[i + 1].gsub!(/[01]/, '0' => 'a', '1' => 'b')
    end
    s[0].gsub!(/#{t[i]}/, t[i + 1])
  end
  puts s[0].delete('x').gsub(/[ab]/, 'a' => '0', 'b' => '1')
end
